lib/wikidata.rb

```diff
@@ -18,17 +18,50 @@ def self.ids_from_pages(lang, titles)
         titles:     sliced.join('|'),
         token_type: false,
       }
-      response = client.action :query, page_args
-      redirected_from = Hash[(response.data['redirects'] || []).map { |h| [h['to'], h['from']] }]
-      response.data['pages'].select { |_k, v| v.key? 'pageprops' }.map do |_k, v|
-        [redirected_from[v['title']] || v['title'], v['pageprops']['wikibase_item']]
-      end
+      TitlesAndIds.new(titles, client.action(:query, page_args)).all
     end
     results = Hash[res.flatten(1)]
     missing = titles - results.keys
     warn "Can't find Wikidata IDs for: #{missing.join(', ')} in #{lang}" if missing.any?
     results
   end
+
+  class TitlesAndIds
```

Comment:
I'm afraid I still don't really understand what this class is or does. I think you need to document this.